Toolkit/ArcGISToolkit/PopupController.swift

```diff
@@ -0,0 +1,347 @@
+        }
+        
+    }
+    
+    @objc private func navigateBackToPopupsFromGeometryEditing(){
+        
+        guard let pvc = self.pvc else{
+            return
+        }
+        
+        editingGeometry = false
+        
+        if let geoViewController = geoViewController, let nc = geoViewController.navigationController{
+            // if there is a navigationController available reset to original buttons
+            geoViewController.navigationItem.rightBarButtonItems = geoViewControllerOriginalRightBarButtonItems
+            geoViewControllerOriginalRightBarButtonItems = nil
+            
+            if useNavigationControllerIfAvailable{
+                nc.pushViewController(pvc, animated: true)
+            }
+            else{
+                geoViewController.present(pvc, animated: true, completion: nil)
+            }
+        }
+        else{
+            geoViewController?.present(pvc, animated: true, completion: nil)
+        }
+    }
+    
+    public func popupsViewController(_ popupsViewController: AGSPopupsViewController, sketchEditorFor popup: AGSPopup) -> AGSSketchEditor? {
+        
+        // give the popupsViewController the sketchEditor
+        
+        if let g = popup.geoElement.geometry{
+            self.sketchEditor.start(with: g)
+        }
+        else if let f = popup.geoElement as? AGSFeature, let ft = f.featureTable as? AGSArcGISFeatureTable{
+            self.sketchEditor.start(with: ft.geometryType)
+        }
+        else{
+            self.sketchEditor.start(with: AGSSketchCreationMode.polygon)
+        }
+        
+        return self.sketchEditor
+    }
+    
+    public func popupsViewController(_ popupsViewController: AGSPopupsViewController, readyToEditGeometryWith sketchEditor: AGSSketchEditor?, for popup: AGSPopup) {
+        // geometry editing has started - show map
+        self.navigateToMapActionForGeometryEditing()
+    }
+    
+    public func popupsViewController(_ popupsViewController: AGSPopupsViewController, didChangeToCurrentPopup popup: AGSPopup) {
+        if let f = popup.geoElement as? AGSArcGISFeature,
+            let ft = f.featureTable as? AGSServiceFeatureTable,
+            let fl = ft.featureLayer{
+            
+            unselectLastSelectedFeature()
+            
+            fl.select(f)
+            self.lastSelectedFeature = f
+            self.lastSelectedFeatureLayer = fl
+        }
+    }
+    
+    public func popupsViewController(_ popupsViewController: AGSPopupsViewController, didDeleteFor popup: AGSPopup) {
+    }
+    
+    public func popupsViewController(_ popupsViewController: AGSPopupsViewController, didFinishEditingFor popup: AGSPopup) {
+        
+        // geometry editing has ended
+        self.sketchEditor.stop()
+        
+        // apply edits for service feature table
+        if let f = popup.geoElement as? AGSArcGISFeature, let ft = f.featureTable as? AGSServiceFeatureTable{
+            ft.applyEdits { (results, error) in
```

Comment:
How about more documentation on the results and potential errors?  The "possible success applying edits..." line begs the question:  "how do I know if it really succeeded?"

Reply:
good question, that will take some thought. Let me look into it. I think IIRC the edit results are nested - thank you